arch: arm: aarch32: add IRQ relay mechanism to ARMv7/8-M

e-rk opened the pull request:
This commit allows the `SW_VECTOR_RELAY` and `SW_VECTOR_RELAY_CLIENT` pair to be
enabled on the ARMv7-M architectures and covers
all additional interrupt vectors.

~~blocked by https://github.com/zephyrproject-rtos/zephyr/pull/26289~~

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

e-rk committed:
arch/arm/aarch32: add IRQ relay mechanism to ARMv7/8-M

This patch allows the `SW_VECTOR_RELAY` and
`SW_VECTOR_RELAY_CLIENT` pair to be
enabled on the ARMv7-M and ARMv8-M architectures
and covers all additional interrupt vectors.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>
Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

e-rk commented on arch/arm/core/aarch32/irq_relay.S:
```diff
@@ -104,5 +104,215 @@ SECTION_FUNC(vector_relay_table, __vector_relay_table)
 	.word __vector_relay_handler
 	.word __vector_relay_handler
+#ifdef CONFIG_ARMV7_M_ARMV8_M_MAINLINE
+	.word __vector_relay_handler
```
Since not all 240 external interrupts are used on all MCUs, I'm wondering if it is possible to reduce the number of defined vectors to conserve flash memory.

nvlsianpu replied:
@anangl Is interrupts number available via DT somehow?

ioannisg replied:
Yes, they are.
@nvlsianpu I am wondering: why do we need to define the full-vector if we are not building for a bootloader, but for a chain-loadable firwmare? That firmware will not use this vector, right? (Well, only the MSP and Reset words are needed).
This dates back to #4536, but can't find the reasoning.

anangl replied:
> @anangl Is interrupts number available via DT somehow?

There is `NUM_IRQS` in Kconfig. It could be used here.

e-rk replied:
`NUM_IRQS` helped here. Thank you @anangl  for suggestion.

e-rk replied:
@ioannisg Indeed, the last binary in the chain does not need the relay vector table. The `relocate_vector_table` in `aarch32/prep_c.c` would have to do nothing in this case.

nvlsianpu replied:
@e-rk  looks like chain-loaded application has build in __vector_relay_handler referenced in __vector_relay_table build which is not used. Am I right?

ioannisg replied:
no @e-rk, I don't think this is correct. The chained binary still needs to set the vector start on the  ram address (normally on 0x20000000), for the bootloader to forward interrupts. But it does not need the relay table.

e-rk replied:
Sorry I phrased it badly. I meant that the last binary *should* not need the vector relay handler, but that's not the case as it is now. Should I address this in this PR?

ioannisg replied:
No need, that's only an enhancement to save some space, but might require Kconfig API changes, so I suggest to ignore for now and get this PR fix only the extension of SW IRQ forwarding in Mainline Cortex-M

ioannisg reviewed (commented):
Looks fine to me.
You need to change the commit message to reflect that this is also done for ARMv8-M architecture.

ioannisg commented on arch/arm/core/aarch32/prep_c.c:
```diff
@@ -37,7 +37,7 @@
 #include <string.h>
 
-#ifdef CONFIG_CPU_CORTEX_M_HAS_VTOR
+#if defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) && !defined(CONFIG_SW_VECTOR_RELAY)
```
This is not good. You cannot skip VTOR initialization on a device that has VTOR. Even if you build a Bootloader, you can't guarantee that this is the first image in boot sequence, so VTOR needs to be determined.

ioannisg replied:
Also, if this is required, particularly, if the relay handler is not at the beginning of flash. There are SoCs that require special sections before the ROM vector table, so VTOR requires to be set.

nvlsianpu replied:
so we want that only bootloader touch VTOR.

nvlsianpu replied:
#ifdef CONFIG_CPU_CORTEX_M_HAS_VTOR

#ifdef CONFIG_XIP
#define VECTOR_ADDRESS ((uintptr_t)_vector_start)
#else
#define VECTOR_ADDRESS CONFIG_SRAM_BASE_ADDRESS
#endif

#else

#if defined(CONFIG_SW_VECTOR_RELAY)
Z_GENERIC_SECTION(.vt_pointer_section) void *_vector_table_pointer;
#endif

#define VECTOR_ADDRESS 0
#endif /* CONFIG_CPU_CORTEX_M_HAS_VTOR */



``` C
#if CONFIG_CPU_CORTEX_M_HAS_VTOR
static inline void relocate_vector_table(void)
#else
void __weak relocate_vector_table(void)
#endif /* CONFIG_CPU_CORTEX_M_HAS_VTOR */
{
#if defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) && !defined(CONFIG_BOOTLOADER_MCUBOOT)
	SCB->VTOR = VECTOR_ADDRESS & SCB_VTOR_TBLOFF_Msk;
	__DSB();
	__ISB();
#endif
#if !defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) || defined(CONFIG_SW_VECTOR_RELAY)
#if defined(CONFIG_XIP) && (CONFIG_FLASH_BASE_ADDRESS != 0) || \
    !defined(CONFIG_XIP) && (CONFIG_SRAM_BASE_ADDRESS != 0)
	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;
	(void)memcpy(VECTOR_ADDRESS, _vector_start, vector_size);
#elif defined(CONFIG_SW_VECTOR_RELAY)
	_vector_table_pointer = _vector_start;
#endif
#endif
}
```
	__DSB();
	__ISB();

#if defined(CONFIG_XIP) && (CONFIG_FLASH_BASE_ADDRESS != 0) || \
    !defined(CONFIG_XIP) && (CONFIG_SRAM_BASE_ADDRESS != 0)
	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;
	(void)memcpy(VECTOR_ADDRESS, _vector_start, vector_size);
#elif defined(CONFIG_SW_VECTOR_RELAY)
	_vector_table_pointer = _vector_start;
#endif
}

e-rk replied:
I have applied the suggestion with one major change. The relay table address is under `__vector_relay_table` variable, therefore `VECTOR_ADDRESS` was set accordingly.

ioannisg reviewed (changes requested):
We need to rework the logic around VTOR initialization.

ioannisg reviewed (commented):
I think this could be a good extension for the SW Interrupt forwarding mechanism. Might have interesting use-cases beyond the Cortex-M Baseline and into the "Mainline" ecosystem.

I'd like a feedback by @d3zd3z on this, possibly also other ARM maintainers (@MaureenHelm , @galak fYI, any objections? I guess we can sandbox this with proper Kconfigs, so default logic does not change.)

nvlsianpu committed:
arch/arm/aarch32: ensured SW IRQ relay modes exclusive

Select either SW_VECTOR_RELAY or SW_VECTOR_RELAY_CLIENT
at the time.

Removed #ifdef-ry in irq_relay.S as SW_VECTOR_RELAY was
refined so it became reserved for the bootloader and it
conditionally includes irq_relay.S for compilation.
See SHA #fde3116f1981cf152aadc2266c66f8687ea9f764

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

nvlsianpu commented:
> I rebased and adapted to mainline changes.
I made the backup of @e-rk previous branch commit history here: https://github.com/nvlsianpu/zephyr/tree/backup/allow-irq-relay-for-armv7-m

nvlsianpu commented:
@ioannisg Please revisit.

ioannisg commented:
> @ioannisg Please revisit.

Yeap, will do this tomorrow morning.

stephanosio commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -267,21 +267,21 @@ config DYNAMIC_DIRECT_INTERRUPTS
-	  This is only
-	  required for Cortex-M0 (or an Armv8-M baseline core) with no hardware
-	  vector table relocation mechanisms or for Cortex-M0+
+	  Normally this is only required for Cortex-M0 (or an Armv8-M baseline core)
+	  with no hardware vector table relocation mechanisms or for Cortex-M0+
 	  (or an Armv8-M baseline core) with no VTOR and no other hardware
 	  relocation table mechanisms.
```
```suggestion
	  Normally this is only required for Cortex-M0(+) and Armv8-M baseline
	  cores with no hardware vector table relocation mechanisms (e.g. VTOR).
```

e-rk replied:
Done.

ioannisg commented on arch/arm/core/aarch32/irq_relay.S:
```diff
@@ -26,10 +26,7 @@
- * is chain-loadable and does not need to forward interrupts.
- */
-#if defined(CONFIG_SW_VECTOR_RELAY) && !defined(CONFIG_BOOTLOADER_MCUBOOT)
+#if defined(CONFIG_SW_VECTOR_RELAY)
```
Is it possible to completely remove this #ifdef and make this conditional in CMakeLists?

ioannisg replied:
actually, i see that irq_relay.S is already compiled-in only when SW_VECTOR_RELAY is set

e-rk replied:
Done.

nvlsianpu replied:
sounds reasonably ^^

ioannisg commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -267,21 +267,18 @@ config DYNAMIC_DIRECT_INTERRUPTS
-	  vector table relocation mechanisms or for Cortex-M0+
-	  (or an Armv8-M baseline core) with no VTOR and no other hardware
-	  relocation table mechanisms.
+	  Normally this is only required for Cortex-M0(+) and Armv8-M baseline
```
Just write "Cortex-M Baseline CPUs", instead of "Cortex-M0(+) and Armv8-M baseline cores"

ioannisg commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -267,21 +267,18 @@ config DYNAMIC_DIRECT_INTERRUPTS
-	  relocation table mechanisms.
+	  Normally this is only required for Cortex-M0(+) and Armv8-M baseline
+	  cores with no hardware vector table relocation mechanisms (e.g. VTOR).
+	  Some specific SW architecture may require this feature despite
```
Last sentence is not really needed. You can add a "not limited to" in the above sentence, after "is only required"

e-rk replied:
Done.

nvlsianpu commented:
Tested on nRF51 and nRF52. Works fine.

ioannisg reviewed (commented):
@e-rk could you please, explain why we need to do sw-vector-relay-client if platform has VTOR?
(i understand the use case with sw-vector-relay, i.e. a bootloader that wants to use sw vector relocation)

ioannisg commented on arch/arm/core/aarch32/prep_c.c:
```diff
@@ -40,43 +40,53 @@
 void __weak relocate_vector_table(void)
+#endif
 {
+#if defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) && !defined(CONFIG_SW_VECTOR_RELAY_CLIENT)
```
It's a bit difficult to see the use-case here: if I have VTOR, why do I need to be a chain-loadable leaf firmware client? I can just make VTOR point to my vector table.

ioannisg replied:
To put it differently: in line 46 you set VECTOR_ADDRESS for the case where HAS_VTOR and RELAY_CLIENT are both set, but this value is not used anywhere.

d3zd3z replied:
I agree with wondering what the reason is to be able to have both VTOR and the relay feature enabled?

ioannisg commented:
Coming back to the original comment: If VTOR exists, it must always be set by the firmware during boot. This still allows you to do SW_VECTOR_RELAY as a bootloader, for non Mainline Cortex-M.

d3zd3z commented on arch/arm/core/aarch32/prep_c.c:
```diff
@@ -40,43 +40,53 @@
+#endif
 {
+#if defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) && !defined(CONFIG_SW_VECTOR_RELAY_CLIENT)
+	SCB->VTOR = VECTOR_ADDRESS & SCB_VTOR_TBLOFF_Msk;
```
I realize this was just moved from the code above, but I don't see how this would work. If the VECTOR_ADDRESS is not aligned according to the mask, it just won't work. Masking off the low bits won't help that (in fact, the hardware will happily do that for you when you write the value).

e-rk replied:
I agree, at least on Cortex-M4, the hardware does it for you.
From Cortex-M4 Devices Generic User Guide
>Table alignment requirements mean that bits[6:0] of the table offset are always zero.

d3zd3z replied:
It isn't as much a matter of whether the mask is there or not, but that if the masked out bits are set, there isn't any hope of the vector table working. The HW doesn't add the VTOR value to the given address, it just uses the upper bits from the mask, and uses the lower bits from the vector. In other words, it doesn't matter whether you mask it or not.

The alignment depends on the particular SoC, based on the number of IRQ vectors used in that device.

It might be nice, at compile time, to detect the table being not at an aligned address.

ioannisg replied:
>The alignment depends on the particular SoC, based on the number of IRQ vectors used in that device.

>It might be nice, at compile time, to detect the table being not at an aligned address.

@d3zd3z this is addressed now in #26800 . In fact this was missing in the code. Now we ensure at compile-time the vector table is properly aligned.

d3zd3z reviewed (commented):
Have you been able to test the configuration with no VTOR (or is that why this can be enabled even without that), and the relay table, with the real vtable in mcuboot, relaying through a table setup by an app? If so, I'll add my approval.

e-rk commented:
@ioannisg Good point. If the Zephyr binary it the only application (aside from the bootloader) that runs on a CPU with VTOR, there is no reason for it to be the sw relay client. It could be used however in legacy SW architectures where the Zephyr application is only a part of a bigger legacy system, however that would be a very rare use case I think. Should I block the CPUs with VTOR from enabling SW_RELAY_CLIENT?

e-rk committed:
arch: arm: aarch32: Always use VTOR when it is available

Zephyr applications will always use the VTOR register when it is
available on the CPU and the register will always be configured
to point to applications vector table during startup.

SW_VECTOR_RELAY_CLIENT is meant to be used only on baseline ARM cores.

SW_VECTOR_RELAY is intended to be used only by the bootloader.
The bootloader may configure the VTOR to point to the relay table
right before chain-loading the application.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

e-rk commented:
@d3zd3z After discussion with @ioannisg and @nvlsianpu, we decided to keep changes in `prep_c.c` to a minimum.
The flow would be as follows:
1. The mcuboot starts
1. Set the VTOR to the actual vector table of the mcuboot in `relocate_vector_table`
2. Do mcuboot processing
3. Before chain-loading the application, the VTOR is set to point to the relay vector table
4. If `CONFIG_BOOT_INTR_VEC_RELOC` is set, `_vector_table_pointer` to point to applications vector table
5. Chain-loaded Zephyr application always relocates the vector using the VTOR, if it is available. Otherwise it sets the `_vector_table_pointer`. It all happens in `relocate_vector_table`.

I have tested it on nRF51 and nRF52.
I'll make the appropriate PR to mcuboot later.

e-rk commented on arch/arm/core/aarch32/cortex_m/relay_vector_table.ld:
```diff
@@ -6,5 +6,3 @@
 KEEP(*(.vector_relay_table))
 KEEP(*(".vector_relay_table.*"))
-KEEP(*(.vector_relay_handler))
-KEEP(*(".vector_relay_handler.*"))
```
I have removed this section and put the `__vector_relay_handler` in `.text`. That way, the two vector tables are one after another and it is more alignment-friendly.

I still have one unresolved problem however, and I'm not sure how to approach this.

The relay vector table gets put at the beginning, so it is always aligned. That is not the case for the real vector table of the application. The alignment is dependent on the number of interrupt handlers.

From Cortex -M4 Devices Generic User Guide

>When setting TBLOFF, you must align the offset to the number of exception entries in the vector
table. The minimum alignment is 32 words, enough for up to 16 interrupts. For more interrupts,
adjust the alignment by rounding up to the next power of two. For example, if you require 21
interrupts, the alignment must be on a 64-word boundary because the required table size is 37
words, and the next power of two is 64.

In `vector_table.ld` how would I have the `exc_vector_table` section be aligned to the nearest power of two depending on its size?

ioannisg replied:
@e-rk can you file an issue about the vector table not being aligned?
I think this is a generic issue that we need to address, regardless of whether we do interrupt forwarding or not.

ioannisg replied:
Nevermind, I posted a PR that takes care of this (needs review, first but you can try it and see it if solves your problem)
https://github.com/zephyrproject-rtos/zephyr/pull/26800

e-rk replied:
Thank you for doing this. It solves my problem. I'll remember now that Kconfig defines can be also used in ld scripts.

ioannisg commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -267,21 +267,17 @@ config DYNAMIC_DIRECT_INTERRUPTS
 
 config SW_VECTOR_RELAY_CLIENT
 	bool "Enable Software Vector Relay (client)"
+	depends on !SW_VECTOR_RELAY && !(CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP || CPU_CORTEX_M_HAS_VTOR)
```
nit: move depends on below default (so the diff disappears)

ioannisg commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -267,21 +267,17 @@ config DYNAMIC_DIRECT_INTERRUPTS
 config SW_VECTOR_RELAY_CLIENT
 	bool "Enable Software Vector Relay (client)"
+	depends on !SW_VECTOR_RELAY && !(CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP || CPU_CORTEX_M_HAS_VTOR)
 	default y if BOOTLOADER_MCUBOOT
```
Since you are here: maybe reword the help text, as follows:

```
Another image has enabled SW_VECTOR_RELAY, and will be forwarding
exceptions and HW interrupts to this image. Enable this option to make
sure the vector table pointer in RAM is set properly by the image upon
initialization.
```
@nvlsianpu agree this is more clear?

e-rk replied:
I have applied the change.

ioannisg commented on arch/arm/core/aarch32/cortex_m/vector_table.S:
```diff
@@ -83,4 +83,3 @@ SECTION_SUBSEC_FUNC(exc_vector_table,_vector_table_section,_vector_table)
 #else
     .word 0
 #endif /* CONFIG_CPU_CORTEX_M_HAS_SYSTICK */
-
```
can you revert this change ?

ioannisg commented on include/linker/linker-defs.h:
```diff
@@ -163,6 +163,10 @@ extern char _image_rodata_size[];
 extern char _vector_start[];
 extern char _vector_end[];
 
+#ifdef CONFIG_SW_VECTOR_RELAY
```
Is this now needed @e-rk ?

e-rk replied:
This address will be used in mcuboot main.c

ioannisg replied:
ok thanks

ioannisg reviewed (commented):
This looks good to me now, @nvlsianpu @e-rk.

In brief, the PR allows a bootloader to use the SW_VECTOR_RELAY feature on Mainline Cortex-M variants.

Should we hold on and merge this after we resolve #26800?
@e-rk will you send up the MCUboot PR as well?

e-rk commented:
@ioannisg Yes, let's wait for #26800.
I'll send the PR to MCUboot tomorrow, as I'm on the move today.

nvlsianpu commented:
^^ sounds good
